{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISM supplemental data \n",
    "- ISM results for all positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import gzip\n",
    "bases = ['A', 'C', 'G', 'T']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from genome import Genome\n",
    "genome = Genome('/Users/jpaggi/Downloads/MPRA/snps/hg19.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# file is result of bedtools intersect -loj\n",
    "introns = {}\n",
    "with open('../preprocessing/introns_to_mercer.tsv') as fp:\n",
    "    for line in fp:\n",
    "        chrom, start, end, _, pos, strand, _, bp = line.split('\\t')[:8]\n",
    "        bp, start, end = int(bp), int(start), int(end)\n",
    "        \n",
    "        three = end if strand == '+' else start  \n",
    "        key = (chrom, three, strand)\n",
    "        \n",
    "        if not 5 < abs(bp - three) < 60:\n",
    "            bp = -1\n",
    "        \n",
    "        if key not in introns: introns[key] = []\n",
    "        if (bp not in introns[key]\n",
    "            and bp != -1): introns[key] += [bp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "L = 70\n",
    "\n",
    "def onehot(seq):\n",
    "    X = np.zeros((len(seq), len(bases)))\n",
    "    for i, char in enumerate(seq):\n",
    "        X[i, bases.index(char)] = 1\n",
    "    return X\n",
    "\n",
    "def encode(introns):\n",
    "    X, y, chroms, coords, strands = [], [], [], [], []\n",
    "    for intron, bps in introns.items():\n",
    "        chrom, three, strand = intron\n",
    "        if strand == '+':\n",
    "            begin, stop = three - L, three\n",
    "        else:\n",
    "            begin, stop = three, three + L\n",
    "\n",
    "        # Get features\n",
    "        seq = genome.get_seq(chrom, begin, stop, strand)\n",
    "        if 'N' in seq: seq = seq.replace('N', 'A')\n",
    "\n",
    "        X += [onehot(seq).reshape(1, 70, 4)]\n",
    "\n",
    "        # Make target\n",
    "        _y = np.zeros((stop - begin,))\n",
    "        for bp in bps:\n",
    "            if strand == '+':\n",
    "                bp = L + bp - three\n",
    "            else:\n",
    "                bp = L - bp + three - 1\n",
    "            _y[bp] = 1\n",
    "        y += [_y]\n",
    "        \n",
    "        chroms += [chrom]\n",
    "        coords += [three]\n",
    "        strands += [strand]\n",
    "    return (np.vstack(X), np.vstack(y),\n",
    "            np.array(chroms), np.array(coords), np.array(strands))\n",
    "\n",
    "(X, y, chroms, coords, strands) = encode(introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from keras.models import load_model\n",
    "model = load_model('../models/2layer.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Write genome-wide ISM to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      "26000\n",
      "27000\n",
      "28000\n",
      "29000\n",
      "30000\n",
      "31000\n",
      "32000\n",
      "33000\n",
      "34000\n",
      "35000\n",
      "36000\n",
      "37000\n",
      "38000\n",
      "39000\n",
      "40000\n",
      "41000\n",
      "42000\n",
      "43000\n",
      "44000\n",
      "45000\n",
      "46000\n",
      "47000\n",
      "48000\n",
      "49000\n",
      "50000\n",
      "51000\n",
      "52000\n",
      "53000\n",
      "54000\n",
      "55000\n",
      "56000\n",
      "57000\n",
      "58000\n",
      "59000\n",
      "60000\n",
      "61000\n",
      "62000\n",
      "63000\n",
      "64000\n",
      "65000\n",
      "66000\n",
      "67000\n",
      "68000\n",
      "69000\n",
      "70000\n",
      "71000\n",
      "72000\n",
      "73000\n",
      "74000\n",
      "75000\n",
      "76000\n",
      "77000\n",
      "78000\n",
      "79000\n",
      "80000\n",
      "81000\n",
      "82000\n",
      "83000\n",
      "84000\n",
      "85000\n",
      "86000\n",
      "87000\n",
      "88000\n",
      "89000\n",
      "90000\n",
      "91000\n",
      "92000\n",
      "93000\n",
      "94000\n",
      "95000\n",
      "96000\n",
      "97000\n",
      "98000\n",
      "99000\n",
      "100000\n",
      "101000\n",
      "102000\n",
      "103000\n",
      "104000\n",
      "105000\n",
      "106000\n",
      "107000\n",
      "108000\n",
      "109000\n",
      "110000\n",
      "111000\n",
      "112000\n",
      "113000\n",
      "114000\n",
      "115000\n",
      "116000\n",
      "117000\n",
      "118000\n",
      "119000\n",
      "120000\n",
      "121000\n",
      "122000\n",
      "123000\n",
      "124000\n",
      "125000\n",
      "126000\n",
      "127000\n",
      "128000\n",
      "129000\n",
      "130000\n",
      "131000\n",
      "132000\n",
      "133000\n",
      "134000\n",
      "135000\n",
      "136000\n",
      "137000\n",
      "138000\n",
      "139000\n",
      "140000\n",
      "141000\n",
      "142000\n",
      "143000\n",
      "144000\n",
      "145000\n",
      "146000\n",
      "147000\n",
      "148000\n",
      "149000\n",
      "150000\n",
      "151000\n",
      "152000\n",
      "153000\n",
      "154000\n",
      "155000\n",
      "156000\n",
      "157000\n",
      "158000\n",
      "159000\n",
      "160000\n",
      "161000\n",
      "162000\n",
      "163000\n",
      "164000\n",
      "165000\n",
      "166000\n",
      "167000\n",
      "168000\n",
      "169000\n",
      "170000\n",
      "171000\n",
      "172000\n",
      "173000\n",
      "174000\n",
      "175000\n",
      "176000\n",
      "177000\n",
      "178000\n",
      "179000\n",
      "180000\n",
      "181000\n",
      "182000\n",
      "183000\n",
      "184000\n",
      "185000\n",
      "186000\n",
      "187000\n",
      "188000\n",
      "189000\n",
      "190000\n",
      "191000\n",
      "192000\n",
      "193000\n",
      "194000\n",
      "195000\n",
      "196000\n",
      "197000\n",
      "198000\n",
      "199000\n",
      "200000\n",
      "201000\n",
      "202000\n",
      "203000\n",
      "204000\n",
      "205000\n",
      "206000\n"
     ]
    }
   ],
   "source": [
    "# Should take about 10 hours...\n",
    "done = 0\n",
    "with gzip.open('../anno/lstm.ism.tsv.gz', 'w') as fp:\n",
    "    fp.write('\\t'.join(['chrom', 'pos', 'ref', 'alt', 'strand', '3pSS', 'd_best_bp', 'd_max_bp'])+'\\n')\n",
    "    for p, seq, chrom, pos, strand in zip(preds, X, chroms, coords, strands):\n",
    "        # Construct alternative sequences\n",
    "        seqs = [] \n",
    "        for i in range(70):\n",
    "            for j in range(4):\n",
    "                if seq[i, j]: continue\n",
    "                s = np.zeros(seq.shape)\n",
    "                s[seq == 1] = 1\n",
    "                s[i] = 0\n",
    "                s[i, j] = 1\n",
    "                seqs += [s]\n",
    "        p = p.flatten()\n",
    "\n",
    "        altpreds = model.predict(np.vstack(seqs).reshape(-1, 70, 4)).reshape(-1, 70)\n",
    "\n",
    "        maxdelta  = np.max(altpreds, axis = 1) - np.max(p)\n",
    "        bestdelta = altpreds[:, np.argmax(p)] - np.max(p)\n",
    "\n",
    "        idx = 0\n",
    "        for i in range(70):\n",
    "            coord = pos + (i-L+1 if strand == '+' else L-i)\n",
    "            ref = bases[np.argmax(seq[i])]\n",
    "            if strand == '-': ref = genome.revcomp(ref)\n",
    "            for j, alt in enumerate(bases):\n",
    "                if seq[i, j]: continue\n",
    "                if strand == '-': alt = genome.revcomp(alt)\n",
    "                fp.write('\\t'.join(map(str, [chrom, coord, ref, alt, strand, pos,\n",
    "                                             bestdelta[i], maxdelta[i]]))+'\\n')\n",
    "                idx += 1\n",
    "        if not done % 1000: print done\n",
    "        done += 1"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
